src/algebra/lie/free.lean

```diff
@@ -0,0 +1,235 @@
+lemma lift_aux_map_add (f : X → L) (a b : lib R X) :
+  lift_aux R f (a + b) = (lift_aux R f a) + (lift_aux R f b) :=
+non_unital_alg_hom.map_add _ a b
+
+lemma lift_aux_map_mul (f : X → L) (a b : lib R X) :
+  lift_aux R f (a * b) = ⁅lift_aux R f a, lift_aux R f b⁆ :=
+non_unital_alg_hom.map_mul _ a b
+
+lemma lift_aux_spec (f : X → L) (a b : lib R X) (h : free_lie_algebra.rel R X a b) :
+  lift_aux R f a = lift_aux R f b :=
+begin
+  induction h,
+  case rel.lie_self : a'
+  { simp only [lift_aux_map_mul, non_unital_alg_hom.map_zero, lie_self], },
+  case rel.leibniz_lie : a' b' c'
+  { simp only [lift_aux_map_mul, lift_aux_map_add, sub_add_cancel, lie_lie], },
+  case rel.smul : t a' b' h₁ h₂
+  { simp only [lift_aux_map_smul, h₂], },
+  case rel.add_right : a' b' c' h₁ h₂
+  { simp only [lift_aux_map_add, h₂], },
+  case rel.mul_left : a' b' c' h₁ h₂
+  { simp only [lift_aux_map_mul, h₂], },
+  case rel.mul_right : a' b' c' h₁ h₂
+  { simp only [lift_aux_map_mul, h₂], },
+end
+
+/-- The quotient map as a `non_unital_alg_hom`. -/
+def mk : non_unital_alg_hom R (lib R X) (free_lie_algebra R X) :=
+{ to_fun    := quot.mk (rel R X),
+  map_smul' := λ t a, rfl,
+  map_zero' := rfl,
+  map_add'  := λ a b, rfl,
+  map_mul'  := λ a b, rfl, }
+
+/-- The functor `X ↦ free_lie_algebra R X` from the category of types to the category of Lie
+algebras over `R` is adjoint to the forgetful functor in the other direction. -/
+def lift : (X → L) ≃ (free_lie_algebra R X →ₗ⁅R⁆ L) :=
+{ to_fun    := λ f,
+    { to_fun    := λ c, quot.lift_on c (lift_aux R f) (lift_aux_spec R f),
+      map_add'  := by { rintros ⟨a⟩ ⟨b⟩, rw ← lift_aux_map_add, refl, },
+      map_smul' := by { rintros t ⟨a⟩, rw ← lift_aux_map_smul, refl, },
+      map_lie'  := by { rintros ⟨a⟩ ⟨b⟩, rw ← lift_aux_map_mul, refl, }, },
+  inv_fun   := λ F, F ∘ (of R),
+  left_inv  := λ f, by { ext x, simp only [lift_aux, of, quot.lift_on_mk, lie_hom.coe_mk,
+    function.comp_app, lib.lift_of_apply], },
+  right_inv := λ F,
+    begin
+      ext ⟨a⟩,
+      let F' := F.to_non_unital_alg_hom.comp (mk R),
+      exact non_unital_alg_hom.congr_fun (lib.lift_comp_of R F') a,
+    end, }
+
+@[simp] lemma lift_symm_apply (F : free_lie_algebra R X →ₗ⁅R⁆ L) : (lift R).symm F = F ∘ (of R) :=
+rfl
+
+variables {R}
+
+@[simp] lemma of_comp_lift (f : X → L) : (lift R f) ∘ (of R) = f :=
+(lift R).left_inv f
+
+@[simp] lemma lift_unique (f : X → L) (g : free_lie_algebra R X →ₗ⁅R⁆ L) :
+  g ∘ (of R) = f ↔ g = lift R f :=
+(lift R).symm_apply_eq
+
+attribute [irreducible] of lift
+
+@[simp] lemma lift_of_apply (f : X → L) (x) : lift R f (of R x) = f x :=
+by rw [← function.comp_app (lift R f) (of R) x, of_comp_lift]
+
+@[simp] lemma lift_comp_of (F : free_lie_algebra R X →ₗ⁅R⁆ L) : lift R (F ∘ (of R)) = F :=
+by { rw ← lift_symm_apply, exact (lift R).apply_symm_apply F, }
+
+/-- See note [partially-applied ext lemmas]. -/
+@[ext] lemma hom_ext {F₁ F₂ : free_lie_algebra R X →ₗ⁅R⁆ L} (h : F₁ ∘ (of R) = F₂ ∘ (of R)) :
```

Comment:
I don't think partially-applied ext lemmas are relevant for normal function composition